06_HelloGraphicsQueue/main.cpp

```diff
@@ -2,80 +2,436 @@
-			// now blit it into the center of the other one
-			cmdbuf->blitImage(,NEAREST);
-			// blit it again to force downsampling with linear filtering
-			cmdbuf->blitImage(,LINEAR);
-			// copy the resulting image to a buffer
-			cmdbuf->copyImageToBuffer();
-			cmdbuf->end();
-
-			// read the buffer back, create an ICPUImage with an adopted buffer over its contents
-			// save the image to PNG,JPG and whatever other non-EXR format we support writing
-
-			return true;
+		// Clear the image to a given colour.
+
+		auto subResourceRange = constructDefaultInitializedSubresourceRange();
+		cmdbuf->clearColorImage(bigImg.get(), IImage::E_LAYOUT::EL_TRANSFER_DST_OPTIMAL, &red, 1u, &subResourceRange);
+		cmdbuf->clearColorImage(smallImg.get(), IImage::E_LAYOUT::EL_TRANSFER_DST_OPTIMAL, &blue, 1u, &subResourceRange);
+
+		IGPUCommandBuffer::SImageMemoryBarrier bigImgClearBarrier = imgClearBarrier;
+		bigImgClearBarrier.barrier.dstAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_WRITE_BIT;
+		bigImgClearBarrier.newLayout = asset::IImage::EL_TRANSFER_DST_OPTIMAL;
+		bigImgClearBarrier.image = bigImg;
+
+		IGPUCommandBuffer::SImageMemoryBarrier smallImgClearBarrier = imgClearBarrier;
+		smallImgClearBarrier.barrier.dstAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_READ_BIT;
+		smallImgClearBarrier.newLayout = asset::IImage::EL_TRANSFER_SRC_OPTIMAL;
+		smallImgClearBarrier.image = smallImg;
+
+		cmdbuf->pipelineBarrier(
+			E_PIPELINE_STAGE_FLAGS::EPSF_TRANSFER_BIT,
+			E_PIPELINE_STAGE_FLAGS::EPSF_TRANSFER_BIT,
+			E_DEPENDENCY_FLAGS::EDF_NONE, 0u, nullptr, 0u, nullptr, 1u, &bigImgClearBarrier
+		);
+
+		cmdbuf->pipelineBarrier(
+			E_PIPELINE_STAGE_FLAGS::EPSF_TRANSFER_BIT,
+			E_PIPELINE_STAGE_FLAGS::EPSF_TRANSFER_BIT,
+			E_DEPENDENCY_FLAGS::EDF_NONE, 0u, nullptr, 0u, nullptr, 1u, &smallImgClearBarrier
+		);
+
+		// Now blit the smallImg into the center of the bigImg.
+		cmdbuf->blitImage(smallImg.get(), IImage::E_LAYOUT::EL_TRANSFER_SRC_OPTIMAL, bigImg.get(), IImage::E_LAYOUT::EL_TRANSFER_DST_OPTIMAL, 1u, &firstImgBlit, ISampler::E_TEXTURE_FILTER::ETF_NEAREST);
+
+		IGPUCommandBuffer::SImageMemoryBarrier bigImgBlitBarrier = constructDefaultInitializedImageBarrier();
+		bigImgBlitBarrier.barrier.srcAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_WRITE_BIT;
+		bigImgBlitBarrier.barrier.dstAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_READ_BIT;
+		bigImgBlitBarrier.oldLayout = asset::IImage::EL_TRANSFER_DST_OPTIMAL;
+		bigImgBlitBarrier.newLayout = asset::IImage::EL_TRANSFER_SRC_OPTIMAL;
+		bigImgBlitBarrier.image = bigImg;
+
+		IGPUCommandBuffer::SImageMemoryBarrier smallImgBlitBarrier = constructDefaultInitializedImageBarrier();
+		smallImgBlitBarrier.barrier.srcAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_READ_BIT;
+		smallImgBlitBarrier.barrier.dstAccessMask = E_ACCESS_FLAGS::EAF_MEMORY_WRITE_BIT;
+		smallImgBlitBarrier.oldLayout = asset::IImage::EL_TRANSFER_SRC_OPTIMAL;
+		smallImgBlitBarrier.newLayout = asset::IImage::EL_TRANSFER_DST_OPTIMAL;
+		smallImgBlitBarrier.image = smallImg;
```

Comment:
btw where's that lambda for joining two halves of pipeline barriers / generating the next one?